pkg/serializer/jsonstream/compressor.go

```diff
@@ -0,0 +1,261 @@
+	New: func() interface{} {
+		return bytes.NewBuffer(make([]byte, 0, maxPayloadSize))
+	},
+}
+
+// compressor is in charge of compressing items for a single payload
+type compressor struct {
+	input               *bytes.Buffer // temporary buffer for data that has not been compressed yet
+	compressed          *bytes.Buffer // output buffer containing the compressed payload
+	zipper              *zlib.Writer
+	header              []byte // json header to print at the beginning of the payload
+	footer              []byte // json footer to append at the end of the payload
+	uncompressedWritten int    // uncompressed bytes written
+	firstItem           bool   // tells if the first item has been written
+	repacks             int    // numbers of time we had to pack this payload
+	maxUnzippedItemSize int
+	maxZippedItemSize   int
+}
+
+func newCompressor(header, footer []byte) (*compressor, error) {
+	c := &compressor{
+		header:              header,
+		footer:              footer,
+		compressed:          bytes.NewBuffer(make([]byte, 0, 1*megaByte)),
+		firstItem:           true,
+		maxUnzippedItemSize: maxPayloadSize - len(footer) - len(header),
+		maxZippedItemSize:   maxUncompressedSize - compression.CompressBound(len(footer)+len(header)),
+	}
+
+	c.input = inputBufferPool.Get().(*bytes.Buffer)
+	c.zipper = zlib.NewWriter(c.compressed)
+	n, err := c.zipper.Write(header)
+	c.uncompressedWritten += n
+
+	return c, err
+}
+
+// checkItemSize checks that the item can fit in a payload. Worst case is used to
+// determine the size of the item after compression meaning we could drop an item
+// that could actually fit after compression. That said it is probably impossible
+// to have a 2MB+ item that is valid for the backend.
+func (c *compressor) checkItemSize(data []byte) bool {
+	return len(data) < c.maxUnzippedItemSize && compression.CompressBound(len(data)) < c.maxZippedItemSize
+}
+
+// hasRoomForItem checks if the current payload has enough room to store the given item
+func (c *compressor) hasRoomForItem(item []byte) bool {
+	uncompressedDataSize := c.input.Len() + len(item)
+	if !c.firstItem {
+		uncompressedDataSize += len(jsonSeparator)
+	}
+	return compression.CompressBound(uncompressedDataSize) <= c.remainingSpace() && c.uncompressedWritten+uncompressedDataSize <= maxUncompressedSize
+}
+
+// pack flushes the temporary uncompressed buffer input to the compression writer
+func (c *compressor) pack() error {
+	expvarsTotalCycles.Add(1)
+	n, err := c.input.WriteTo(c.zipper)
+	if err != nil {
+		return err
+	}
+	c.uncompressedWritten += int(n)
+	c.zipper.Flush()
+	c.input.Reset()
+	return nil
+}
+
+// addItem will try to add the given item
+func (c *compressor) addItem(data []byte) error {
+	// check item size sanity
+	if !c.checkItemSize(data) {
+		return errTooBig
+	}
+	// check max repack cycles
+	if c.repacks >= maxRepacks {
+		return errPayloadFull
+	}
+
+	if !c.hasRoomForItem(data) {
+		if c.input.Len() == 0 {
+			return errPayloadFull
```

Comment:
Why return PayloadFull if the input is empty ?

Reply:
Input is the temporary buffer holding uncompressed data. If there is no room in the current payload being built to add the item and there is no uncompressed data pending compression it means the item has no way to be added to the current payload.

I can explain you the global logic on a virtual whiteboard if you want